gcp/gcp.go

```diff
@@ -399,8 +441,239 @@ func (p gcpProvider) TerminateInstances(instances *types.InstanceContainer) []er
+			if err != nil {
+				errChan <- err
+			}
+			resultChan <- err == nil
+		}(i, zone)
+	}
+
+	if database != "" {
+		wg.Add(1)
+
+		go func(dbName string) {
+			defer wg.Done()
+
+			log.Debugf("[GCP] Terminating db %s", dbName)
+
+			err := p.doAndPollSqlCall(p.sqlClient.Instances.Delete(p.projectID, dbName))
+			if err != nil {
+				errChan <- err
+			}
+			resultChan <- err == nil
+		}(database)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	for success := range resultChan {
+		if !success {
+			return false
+		}
+	}
+	return true
+}
+
+func (p gcpProvider) terminateNetworkRelatedResourcesInStack(stack *types.Stack, errChan chan error) {
+	ips := strings.Split(stack.Metadata["externalIps"], ",")
+	firewall := stack.Metadata["firewall"]
+	subnets := strings.Split(stack.Metadata["subnets"], ",")
+	network := stack.Metadata["network"]
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(ips))
+
+	for _, i := range ips {
+		go func(ip string, region string) {
+			defer wg.Done()
+
+			log.Debugf("[GCP] Terminating ip %s", ip)
+
+			if _, err := p.computeClient.Addresses.Delete(p.projectID, region, ip).Do(); err != nil {
+				errChan <- err
+			}
+		}(i, stack.Region)
+	}
+
+	if network != "" {
+		wg.Add(1)
+
+		go func(networkName string, firewallName string, subnetNames []string) {
+			defer wg.Done()
+
+			preNetworkWg := sync.WaitGroup{}
+			preNetworkWg.Add(1 + len(subnetNames))
+
+			go func(firewallName string) {
+				defer preNetworkWg.Done()
+
+				log.Debugf("[GCP] Terminating firewall %s", firewallName)
+
+				err := p.doAndPollComputeCall(p.computeClient.Firewalls.Delete(p.projectID, firewallName))
+				if err != nil {
+					errChan <- err
+					log.Infof("[GCP] Failed to delete firewall %s, skipping network delete", firewallName)
+					return
+				}
+			}(firewallName)
+
+			for _, s := range subnetNames {
+				go func(subnetName string, region string) {
+					defer preNetworkWg.Done()
+
+					log.Debugf("[GCP] Terminating subnet %s", subnetName)
+
+					err := p.doAndPollComputeCall(p.computeClient.Subnetworks.Delete(p.projectID, region, subnetName))
+					if err != nil {
+						errChan <- err
+					}
+				}(s, stack.Region)
+			}
+
+			preNetworkWg.Wait()
+
+			log.Debugf("[GCP] Terminating network %s", networkName)
+
+			err := p.doAndPollComputeCall(p.computeClient.Networks.Delete(p.projectID, networkName))
+			if err != nil {
+				errChan <- err
+			}
+		}(network, firewall, subnets)
+	}
+
+	wg.Wait()
+}
+
+type SqlCall interface {
+	Do(opts ...googleapi.CallOption) (*sqladmin.Operation, error)
+}
+
+func (p gcpProvider) doAndPollSqlCall(call SqlCall) error {
+	operation, err := call.Do()
+	if err != nil {
+		return err
+	}
+
+	for {
+		time.Sleep(3 * time.Second)
+		log.Debugf("[GCP] Polling sql operation %s", operation.Name)
+		operation, err := p.sqlClient.Operations.Get(p.projectID, operation.Name).Do()
+
+		if err != nil {
+			return err
+		} else if operation.Status == "DONE" {
+			if operation.Error != nil {
```

Comment:
I suggest to check `operation.Error == nil` with an early return

Reply:
Done